### ISSUE TITLE
dynamic runtime environment variables

### ISSUE THREAD

**Issue by janesser** [nuxt/nuxt.js#5100]
While striving for immutable build artifacts we came across some possible improvement points.

First let me explain the motivation behind "immutable build artifact". Alongside of the stages like e,g, DEV, TEST, QA, UAT, PROD the configuration changes, as an example nuxt-axios API_URL and API_BROWSER_URL. Ideally this can be injected by env vars, which pretty much fits the kubernetes way of dealing with configuration.

Presently a work-around can be achieved with help of https://github.com/samtgarson/nuxt-env, which does a decent job in providing needed injection mechanism. The down-side is that no nuxt-plugin by default goes for this.$env, which drives one in a lot of customizing of plugins. Not to mention the broken upgrade ease.

Can you imagine a way to support this natively in some of the next nuxt releases?

<div align="right"><sub><em>This question is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c8719">#c8719</a>)</em></sub></div>

**Comment by pi0** [comment]
Hi @janesser. I've transferred issue from RFC to here because it is more an enhancement discussion.

Values passed to the `env` section of `nuxt.config` will be _replaced_ by a webpack plugin during the build. Would you please elaborate more what you expect from storing it in build artifact?

**Comment by vhf** [comment]
I also ran into this problem, here is one of my example case:

I'm using nuxtjs auth module with some OAuth API. Here is what the config looks like: https://auth.nuxtjs.org/reference/providers/github , in `nuxt.config.js`:

```js
auth: {
  strategies: {
      github: {
        client_id: '...',
```

I'd like to build my app once and deploy it many times with different configurations. This cannot be done because `client_id: process.env.FOO_CLIENT_ID` will be replaced at build time and the build-time value will be written in the build artifacts.

Here you couldn't build your app in a container and deploy it with different `FOO_CLIENT_ID` values, you would have to build the app each time you start a container.

**Comment by aldarund** [comment]
@pi0 its more about have runtime env vars like in the module that is linked. E.g. build once deploy to different environments same artifact, isntead of building each time for each env ( e.g. staging, test, production etc )

**Comment by pi0** [comment]
I see the point. And auth and axios are probably best examples for such need. However, this is a duplicate topic and issue was discussed here is a short brief of status:

- 1. `process.env.*` from **server side** is runtime configurable. axios module supports it for example for `API_URL`
- 2. `process.env.*` from client-side when running in SSR mode, is only changeable if we somehow pass it to `nuxtState`. The simplest way would be using a store module.
- 3.`process.env.*` from client-side when running in SPA or Static generated mode, is only changeable by doing an API call.

Having built-in support for case (2) could be possible but potentially opens a way to critical security holes.

**Comment by janesser** [comment]
@pi0 can you link the original discussion that you refer too, please?

@vhf @aldarund thanks for supporting the need.

regarding the security, i am confident that docker envvars are sufficiently safe.
having something in javascript either hard-coded or dynamically refreshed is also in the same security class. Maybe i miss some attack vectors.

i felt like you hit the sensible point where it came to the way that webpack4 build routine is instrumented. Probably the present behaviour is at least partially inherited.

regarding possible solutions, did you check nuxt-env? by my experience it covers SSR plus non-SSR in a consistent way, not sure how it is solved internally. //cc @samtgarson

ill try to collect some insight on webpack4 and eventually read through the internals of nuxt-env to give you some additional pointers.

**Comment by samtgarson** [comment]
@janesser in the meantime open an issue for any confusing things you're hitting—I picked `$env` as it seemed appropriate but we can look at changing it or allowing it to be dynamic pretty easily.

@pi0 all I do is use a middleware to add the `process.env` values onto the `req` object, then a plugin which uses `nuxtState` to read from `req` and inject them onto the context.

Would be great to have this natively, it's currently impossible to use Nuxt in any kind of containerised, multi-environment setup with the existing native setup, or at least anywhere where the production environment could vary from the build environment.

**Comment by samtgarson** [comment]
(non SSR and SPA are still scenarios that I don't think `nuxt-env` covers _super_ smoothly, but it's OK...)

**Comment by pi0** [comment]
Mainly discussed in #618.

We are aware of @samtgarson. That's implementing (2) method which would be useful for users that want it. It is more than qualified to be a nuxt-community project :+1: 

About security matter. Docker can't prevent one accidentally passing `client_secrect` to the user's browser window. It is the user's responsibility to be aware and ensure that.

/cc @manniL @aldarund

**Comment by janesser** [comment]
About secrets retention, check this one: https://github.com/samtgarson/nuxt-env/issues/9

To wrap up, is there a way to make process.env as consumed by default from nuxt-plugins in the same way as process.$env?

**Comment by samtgarson** [comment]
regarding _secret_ env vars: `nuxt-env` supports these as of [this PR](https://github.com/samtgarson/nuxt-env/pull/17) but as @pi0 says, it's still up to the developer to ensure these don't get passed to the client.

Would be keen to help any efforts of integrating this or moving it over to nuxt-community, let me know if either of those are desired.

**Comment by axetroy** [comment]
This is my solution:

set dynamic runtime environment variables in `store`

so we can get it in client side and server side by `store.state.env`.

```js
// store/index.js

export const state = () => ({
  env: 'development'
})

export const actions {
  // run in every SSR
  async nuxtServerInit (store, context) {
    // read runtime environment everytimes and set to store
    const env = process.env.NODE_ENV || 'development'
    store.commit('SET_ENV', env)
  }
}

export const mutations = {
  SET_ENV (state, env) {
    state.env = env
  }
}
```

**Comment by janesser** [comment]
@pi0 @samtgarson what's the next step here?

**Comment by vhf** [comment]
> I see the point. And auth and axios are probably best examples for such need.

Definitely. The ability to configure auth, axios and apollo at runtime would be immensely helpful to people who wish to build once and deploy many times.

I'd be happy to contribute if the core team agrees on what the right approach should be.

**Comment by fvigotti** [comment]
@axetroy I may be missing something.. but how can your solution be related to current topic ? the process.env.NODE_ENV  gets replaced on build-time with a STATIC-VALUE hardcoded, so the final .nuxt/ directory you will have something like 
    `const env = 'development' || 'development'`
this isn't super-useful with the current issue about **runtime** variables injection, in your solution the `process.env.NODE_ENV` isn't even ridden at runtime.. or I'm missing something?
what is the `ìì - store` solution that @pi0  cited ? 

my 2 cents are that this is a serious issue, I've lost 3 hours trying to figure out why my local dev env works and my deploy env didn't , the problem was that there was two different type of injnected variables because built interpolated them from different build environment,
I think that saying that there is no official way to use env variables at runtime but that https://nuxtjs.org/api/configuration-env/ works only with env variables at build-time  must be written at least here `https://nuxtjs.org/api/configuration-env/` with bold characters untill resolved.. and maybe  propose this module `https://github.com/samtgarson/nuxt-env` as a solution ? anyway having someting so hardly testable, so important (that is used to pass auth / tokens / configuration to the application ) on a non-official module (with all the eventual future compatibility problem, version-pairing, .. )  is a sign of project immaturity and should be incorporated in the main app asap.

**Comment by samtgarson** [comment]
@pi0 I'm happy to move the module over to nuxt community if that would help the situation, let me know

**Comment by manniL** [comment]
@samtgarson Let's discuss that on https://discord.nuxtjs.org/ :relaxed:

**Comment by rmeissn** [comment]
Just to give some idea of a possible solution:
We had the same issue in my old project and we splitted the webpack build step into two steps. One, that builds most parts of the application (e.g. executed by CI/CD) and one smaller step that updates a file with predefined env vars before starting the application. See scripts `build:nostart` and `start` in [package.json](https://github.com/slidewiki/slidewiki-platform/blob/ad29127b9df4c720be799943790d50e45c372b6d/package.json#L30)

So the basic steps for a docker deployment (on container execution) are:
1. An entrypoint script is executed that uses (in our case) `envsubst` and a template file to create a config file, containing predefined env vars (so they are hardcoded to the resulting json file)
2. Run a limited webpack build step in order to include this newly created file into the already existing distribution
3. Execute the distribution (that is now using the env vars used on container startup)

I can perfectly image that these steps may be included into `nuxt start` command.

**Comment by tomsaleeba** [comment]
I have (had) a need for this too and I use nuxt-env (excellent module) where I can but there are places that nuxt-env can't help like configuring your API URL for the nuxt-proxy module in the `nuxt.config.js` file.

I run in Docker so I decided to get a bit hacky and just use placeholders during the docker build:
```yaml
# Dockerfile
...some stuff

RUN \
  export KEYCLOAK_CLIENT_ID='%%KEYCLOAK_CLIENT_ID%%' && \
  export KEYCLOAK_HOST='%%KEYCLOAK_HOST%%' && \
  export KEYCLOAK_REALM='%%KEYCLOAK_REALM%%' && \
  export API_BASE_URL='%%API_BASE_URL%%' && \
  yarn build

... more stuff
```

Then in the `entrypoint.sh` for the container, just parse the build JS files:
```bash
# entrypoint.sh
find .nuxt/ \
  -type f \
  -name '*.js' \
  -exec sed -i "s+%%KEYCLOAK_CLIENT_ID%%+${KEYCLOAK_CLIENT_ID:?}+g" '{}' \; \
  -exec sed -i "s+%%KEYCLOAK_HOST%%+${KEYCLOAK_HOST:?}+g" '{}' \; \
  -exec sed -i "s+%%KEYCLOAK_REALM%%+${KEYCLOAK_REALM:?}+g" '{}' \; \
  -exec sed -i "s+%%API_BASE_URL%%+${API_BASE_URL:?}+g" '{}' \;

exec yarn start
```

It's not glamorous but it gets the job done. It works for local dev (`yarn dev`) too because you can pass the env var directly. The downside is that you have to update the docker files when you need a new var but I can live with that. You could always get fancy and scan the `env` in the entrypoint for a certain prefix and then update all the matching env vars, but that's gold plating for me right now.

**Comment by vhf** [comment]
This is still an important issue for those who'd like to build easily deployable products with Nuxt. It would be good if stalebot didn't close it IMO.

**Comment by janesser** [comment]
Any news on integration of nuxt-env? @samtgarson @pi0

**Comment by samtgarson** [comment]
Waiting to hear from @pi0 or @manniL on how they want this to be taken forward 🙃

**Comment by janesser** [comment]
@pi0 stale or pending?

**Comment by ScreamZ** [comment]
Guys at nuxt, why just not enable a mode where process.env.X remain like this even after build.

This match https://12factor.net/config requirements.
I can't use my kubernetes clusters env injection currently because my env variables are built-in.

*My workaround :* Having a docker image which build on start with a pm2 inside which handle failure and downtimes.

Not really cool for auto-scaling.

**Comment by cheposki** [comment]
i have this problem 
process.env.* from client-side when running in SPA or Static generated mode, is only changeable by doing an API call.
i have to do the build then do the deploy in other enviroments, i cant pass variables to my build

**Comment by SkyaTura** [comment]
I have the same problem, and we also can't pass variables to the build time.

Any updated on this issue?

**Comment by janesser** [comment]
I guess it is time for a leap forward. Somebody volunteering to propose a PR? It doesn't need to be @samtgarson.

Supporters will be found easily, given the recent calls.

**Comment by samtgarson** [comment]
I think it would be good to agree a strategy for the API of this feature before jumping into a PR.

We'd have to agree how it would behave given that it directly conflicts with the existing `env` config attribute.

We could just add a second attribute like `runtimeEnv`, but my suggestion would be to introduce a new format for values in the `env` config to take an object:

```js
export default {
  env: {
    // baked
    baseUrl: process.env.BASE_URL,
    // baked with new syntax
    anotherApiKey: { value: process.env.ANOTHER_API_KEY, runtime: false },
    // fried
    apiKey: { value: process.env.API_KEY, runtime: true },
    // fried with async function?
    someValue: { async value () { return fetchEnvValue() }, runtime: true }
  }
}
```

This would allow existing apps to remain unchanged (the current behaviour will remain as the default) but would allow us to define attributes (and even functions?) which will be evaluated at runtime, leaving extensibility for plugins and asynchronous fetching environment from filesystem, etc.

I haven't taken a look at the underlying code so not sure the impact of this, but considering it's touch webpack build process as well as runtime server and client, I imagine it's not insignificant. Unfortunately I don't have time to kick this off but would happily support with picking up bits and reviewing code if someone else has.

**Comment by SkyaTura** [comment]
I don't know well about Nuxt's build process, but I guess if we use "process.env.FOO" on the new syntax it may not be the best solution, since it would still be replaced by Webpack.

I thought something like your suggestion, but a little different:
```js
export default {
  env: {
    // baked
    baseUrl: process.env.BASE_URL,
    // baked with new syntax
    anotherApiKey: { key: 'ANOTHER_API_KEY', runtime: false },
    // fried
    apiKey: { key: 'API_KEY', runtime: true },
    // fried with async function?
    someValue: { async value () { return fetchEnvValue() }, runtime: true },
  }
}
```

**Comment by samtgarson** [comment]
That’s a good point. From what I know, `nuxt.config.js` is run at build and at runtime, so depending on how it’s inplemented we should be able to evaluate the `process.env.FOO` at both points, but yeah depends on the internals. 

Personally I find providing a value more intuitive than a key, and you can always provide a function that will be executed at runtime to be sure. 

Definitely open to either though!

**Comment by SkyaTura** [comment]
I vote to use both `{ key: 'FOO' }` and `{ value: process.env.FOO }` sintaxes 👀

The possibility of using functions are good as well. In fact, one of my first tries on this issue was passing a function. ~(sadly, didn't work)~

**Comment by SkyaTura** [comment]
So, I was thinking on this subject, and here is my strach of proposal:

### Old syntax
```js
export default {
  env: {
    baseUrl: process.env.BASE_URL || 'http://localhost:3000'
  }
}
```

### New syntax
```js
export default {
  env: {
    oldWay: process.env.OLD_WAY, // behaves as always, for backwards compatibility
    baseUrl: {
       runtime: true,
       key: 'BASE_URL',
       value: 'http://localhost:4000/api',
    },
  },
}
```
| Property | Type | Default | Description |
| - | :-: | :-: | - |
| runtime | Boolean | false | If **true**, the variables won't be replaced by static values, allowing different values during runtime. |
| key | String | | Used to automatically get a value from the environment. <br/> Same as `process.env.${key}` |
| value | String <br/> Function <br/> Promise | | Used to define the value of the env. <br/> If **key** is set, this behaves as a fallback in case of the value is undefined in the environment. <br/> If used as Function/Promise, will consider the result value. |

### Examples
```js
export default {
  env: {
    // Old behavior maintains the same
    baseUrl: process.env.BASE_URL || 'http://localhost:4000/api',
    // New syntax, baked, automatic with fallback value
    baseUrl: {
       key: 'BASE_URL',
       value: 'http://localhost:4000/api',
       // runtime: false, // This is optional in this case
    },
    // New syntax, runtime, manual value
    {
       runtime: true,
       value: process.env.BASE_URL || 'http://localhost:4000/api',
    },
    // New syntax, runtime, automatic with fallback value
    {
       runtime: true,
       key: 'BASE_URL',
       value: 'http://localhost:4000/api', // Fallback value
    },
  },
}
```

**Comment by simplesmiler** [comment]
@pi, I'd like to challenge the assumption that runtime env vars in SPA mode require an API call.

Even in SPA mode, the HTML is rendered in runtime, so runtime env vars can be communicated through a `<meta>` tag. In fact, I consider this to be a preferable method (as compared to passing them via `nuxtState`), because this method will work uniformly in most cases (universal/spa, dev/build+start).

Arguably, this method will also work in the `nuxt generate` case, because it opens a way to change env variables after build (e.g. a `<script src="/env.js">` that modifies the env meta tag before nuxt sees it).

I've made a proof of concept for this method ([1](https://github.com/simplesmiler/nuxt-env-poc/blob/master/spa/modules/nuxt-env-poc/index.js#L9), [2](https://github.com/simplesmiler/nuxt-env-poc/blob/master/spa/modules/nuxt-env-poc/plugin.js#L4)). Currently there is an issue with `nuxt start` in SSR mode (in this particular case nuxt does not seem to respect the modified `options.head.meta`, and instead uses a snapshot of how it looked like during `nuxt build`, which feels like a bug). But an easy workaround is to fallback to `nuxtState` in this particular case ([3](https://github.com/simplesmiler/nuxt-env-poc/blob/master/universal/modules/nuxt-env-poc/plugin.js#L13)) until the issue is resolved.

**Comment by simplesmiler** [comment]
Another important topic is dead code elimination with `process.env`. It is impossible to eliminate dead code based on a runtime env var (assuming we don't want to do a second round of compilation upon `nuxt start`). Thus, using `process.env` for accessing runtime variables would be misleading. I suggest we should keep runtime variables just to `context.env`.

**Comment by janesser** [comment]
@simplesmiler i would appreciate to see your poc.

as stated before nuxt.config.js is handled by webpack,
and also the process.env things should be available on both sides SSR and CSR.
that meta-tag env.js how would it apply to SSR side?

regarding process.env to me it was the opposite way round, i was surprised that switching process.env after build didn't have any effect.

Syntax-wise proposal from @SkyaTura looks excellent to me.
The missing piece of the puzzle is where to put it in the nuxt.config.js processing.

Cheers!

**Comment by simplesmiler** [comment]
@janesser don't get your hopes up :) My approach is not that different to `nuxt-env`, except for that I use a different transport to deliver values to the client, which allows for it to work in spa mode as well as in universal mode. With my approach `process.env` won't be dynamic, only `context.env` will.

I'll try to publish it today as a npm module (using similar syntax to `nuxt-env`, but less features), but in the mean time you can check the code [here](https://github.com/simplesmiler/nuxt-env-poc/tree/master/universal/modules/nuxt-env-poc).

**Comment by simplesmiler** [comment]
I've published proof of concept for communicating runtime env vars without API call as `nuxt-env-shim`. It's based it on @samtgarson's `nuxt-env` and initially I planned to make a PR, but the changes got so significant, that I resorted to publishing a new module.

**Comment by SkyaTura** [comment]
@simplesmiler is it possible to use this with modules like @nuxtjs/axios?

**Comment by simplesmiler** [comment]
@SkyaTura the short answer is "yes, it is possible".

The long answer is "yes, but there might be limitations, and the exact details of making it work depend on the module in question".

Here are the details for the axios module:
1. If you are sending requests to `/` and using the proxy module along with the axios module, then you don't need the shim. Nothing in the proxy module gets baked into the bundle, so using `process.env` for it's configuration already gives you runtime env vars.
2. Otherwise, instead of configuring axios from the `nuxt.config.js` you have to configure it in runtime from a plugin. Then you include the shim module before the axios module, and you're good to go.

I've updated [nuxt-env-shim](https://github.com/simplesmiler/nuxt-env-shim) with `shimModules` option which automatically instruct axios about runtime env vars (`API_HOST`, `API_PORT`, `API_PREFIX`, `API_URL` and `API_URL_BROWSER`).

**Comment by janesser** [comment]
@simplesmiler looks promising. ill give it a try next days. do you have like a sample-project? I would be trying to test from meta-tags-rerender to env-variables re-injected etc.? CC @samtgarson will you have a look?

**Comment by blowsie** [comment]
Is there a plan to implement something into the core?

**Comment by samtgarson** [comment]
I believe that is the point of this issue 🤞

@simplesmiler I like the way you've used the meta to make this work with SPA apps 👍

But I don't believe the idea of shims is worth pursuing and even could be irresponsible to publish, as it's very fragile in the way it's tied to other modules, and will lead to confusion as to where bugs are occurring, etc. 

I believe we're better off helping @pi0, @manniL and the Nuxt core team implement this into core if they'd like.

**Comment by blowsie** [comment]
The reality of using nuxt axios in a docker container, currently using `nuxt-env`

```
export default function({ $axios, app }) {
  if (process.server) {
    $axios.defaults.baseURL = `http://localhost:${app.$env.PORT || '5000'}`
  } else {
    $axios.defaults.baseURL = app.$env.EP_BASE_URL || `http://localhost:${app.$env.PORT || '5000'}`
  }
}
```

**Comment by samtgarson** [comment]
That seems to me to be an issue with the axios module though, not even with Nuxt itself—I haven't tested myself but if the way the module is written causes the options to be baked in then that would most likely need to change even if we get dynamic environment vars in Nuxt core.

I don't use that axios module in production, has anyone raised an issue with them?

**Comment by blowsie** [comment]
I think its more of an issue with `nuxt-axios` attempting to share the the same baseUrl for both client and server doesnt always make sense, for dns reasons.

**Comment by slavanossar** [comment]
Can someone shed some light on how the proposed solutions for extending the `env` option in `nuxt.config.js` would cater to modules making use of runtime environment variables via their own options?

As an example, this is how the [apollo-module](https://github.com/nuxt-community/apollo-module) works:

1. `apollo` option from `nuxt.config.js` is [passed into module on `build`](https://github.com/nuxt-community/apollo-module/blob/master/lib/module.js#L9)

1. [Template plugin](https://nuxtjs.org/guide/modules/#template-plugins) is rendered, passing in options
https://github.com/nuxt-community/apollo-module/blob/master/lib/module.js#L51-L56

1. Option values are used throughout the [plugin's template](https://github.com/nuxt-community/apollo-module/blob/master/lib/templates/plugin.js)

I'm fairly new to Nuxt, so maybe I'm not understanding the proposals correctly (or conflating two separate issues)?

Some follow-up questions:

- Can modules be configured to compile template plugins at runtime?
- How would modules be able to use their own options object while evaluating any values defined by environment variables at runtime?

**Comment by samtgarson** [comment]
@slavanossar Unfortunately from what I can tell that’s up to the modules themselves to make that change.

Nuxt _could_ provide some helpers or system to make that easier for them (e.g. something to allow easy access to the runtime env from inside the module container) but the discussion hasn’t progressed to that. 

Fundamentally we’re still lacking direction from the Nuxt team on how they think all this should be handled.

**Comment by slavanossar** [comment]
@samtgarson Thanks for the answer, that clears it up for me.

I can definitely appreciate that this proposal is still in its infancy, but I think it's worth considering how modules would leverage something similar as there is a need for it too. Maybe the solution will fall out of solving the simpler problem of accessing env vars at runtime?

Hopefully the Nuxt team can appreciate the need for this functionality, I've sunk way too much time into investigating this for my current project 🙃

**Comment by nachogarcia** [comment]
Same issue here. I'm usign dotenv and nuxt-env as follows for my app to have some default env variables and be able to modify them with docker.
```
const { parsed: parsedEnv } = require('dotenv').config()
[...]
  modules: [
    ['nuxt-env', {
      keys: Object.entries(parsedEnv).map(([ key, value ]) => ({ key, default: value }))
    }],
```
But this doesn't work for env variables passed to modules or configs in nuxt.config.js such as
```
  googleAnalytics: {
    id: process.env.GOOGLE_ANALYTICS_ID
  },
[...]
    publicPath: process.env.WEB_URL + '/_nuxt/'
```
IMO this is a completely necessary feature to allow twelve-factor apps, and would need to be 100% supported by default instead of having half support using 2 modules.

**Comment by SkyaTura** [comment]
I'd like to know what is needed to get an official solution for this issue, and how can we help with it?

**Comment by samtgarson** [comment]
I'm also waiting to hear from @manniL and @pi0 to see if this is a prioritised issue or what the path would be to getting it into Nuxt core.

However, even this would not fix the issue with the third-party modules like nuxt-axios and GA—due to the way they are built, they would still bake the config into the bundle at build time, even with a change to Nuxt. They would have to all update to be able to take advantage of this feature.

There are workarounds for most of these—for example in axios you can set the baseUrl at runtime in a plugin using nuxt-env, instead of setting it in nuxt.config.js (e.g. `app.$axios.defaults.baseUrl = app.$env.BASE_URL`)

**Comment by NickBolles** [comment]
would a mechanism to cache off the config that nuxt was built with and then compare that on startup and if it's different re-build the app help fix this? currently with my docker setup I just have it re build each time I start the container if I rely on env vars.

**Comment by nachogarcia** [comment]
IMO docker images should be stateless, so I push to the image the node modules, nuxt config and dist folder. This achieves a fast startup. Not sure if it would work for me.

**Comment by cardaddy** [comment]
I'm trying to build my application with some args from my docker build and I'm not seeing them get populated within my auth module. However the env variable appears to work fine outside of nuxt.config.js. If I use dotenv, auth module is capable of reading the values, however I can't seem to overwrite them with docker. What am I doing wrong?

    strategies: {
      company: {
        _scheme: 'oauth2',
        authorization_endpoint: process.env.BASE_URL + '/auth/oauth/authorize',
        access_token_endpoint: process.env.BASE_URL + '/auth/oauth/token',
        userinfo_endpoint: process.env.BASE_URL + '/auth/users/me',

**Comment by samtgarson** [comment]
@cardaddy I think you need to open that issue with the auth module directly, as I believe it's them that are baking the config into the build output (I think it's in [this line](https://github.com/nuxt-community/auth-module/blob/dev/lib/module/plugin.js#L21)).

You can try using [nuxt-oauth](https://github.com/SohoHouse/nuxt-oauth) if you're just after a basic OAuth integration, we use that in production and it works with docker.

**Comment by ZinkNotTheMetal** [comment]
> app.$axios.defaults.baseUrl = app.$env.BASE_URL

Do you have an example, I just started with NUXT and I'm confused where to set this

**Comment by cardaddy** [comment]
@ZinkNotTheMetal I'm setting the value in nuxt.config.js

  axios: {
    baseURL: process.env.BASE_URL,
    credentials: true
  },

**Comment by ZinkNotTheMetal** [comment]
@cardaddy 
I did that, but these are still build time variables right? I'm trying to pass in an OpenShift (docker) variable into the nuxt.config.js and I feel like I've spent too long trying to do this....

**Comment by danielroe** [comment]
@ZinkNotTheMetal You'll want to do it in a plugin.

**Comment by cardaddy** [comment]
Yes, these are not runtime variables and must be set at build time. Perhaps one of the nuxt guys will be able to shed some light on it, but I struggled to set different base_urls passing in environment variables and still having a default for local dev. I ended up setting up nuxt dotenv and ultimately setting my .env file during my docker build process. I think this is an area that needs a lot of improvement.

**Comment by ZinkNotTheMetal** [comment]
Yes, I have spent the better part of an afternoon trying everything under the sun to get runtime variables into nuxt.   
I don't see how this doesn't break the build once - deploy many times paradigm that javascript seems to strive for. Hopefully this feature will come soon. Thanks for the post and the help

**Comment by cardaddy** [comment]
Yes it does seem to break that paradigm and I find it to be quite frustrating, but hopefully in time it will be resolved. 

If you end up using the nuxt auth module, you'll find it wont be able to read the standard nuxt environment variables in nuxt.config.js auth strategies, you'll need dotenv to get it to work.  I'm too looking for a better solution.

**Comment by samtgarson** [comment]
@ZinkNotTheMetal A plugin works for setting axios defaults using environment variables, although we've had other issues with that module and have switched back to vanilla axios. 

It would be great to hear from nuxt team on this regarding strategy, but the issue is with axios and Auth module among others. 

At Soho House we use nuxt in prod with docker images in kubernetes, and use regular axios (without the nuxt module), nuxt-oauth and other modules along with nuxt-env to handle environment variables just fine—its only an issue with modules that template options into build artefacts.  

(PS I'm sure those modules would accept PRs to fix this 🙃)

**Comment by cardaddy** [comment]
@samtgarson You should create a blog illustrating your build and how you got everything to work. Getting all these modules to work together has been nightmare.

**Comment by superbiche** [comment]
I'm also deploying on Kubernetes, using GitLab's auto review feature.  
After spending the afternoon trying to find a good workaround, I ended up re-building the app before running `yarn start`. So I'm bundling the app in the image during the build, and redo the bundle before it starts.  
Not great for auto scaling or reducing my cluster's production worker load.

**Comment by jslegers** [comment]
I'm having the same problem when trying to configure the `Auth` module to use [`Auth0`](https://auth0.com/) as an authentication service.

The `Nuxt` build is supposed to be the same for our `dev` environment and our `staging` environment, but both environments use a different database and different `auth0` endpoints.

I can use `env` variables to set the  `auth0` endpoints in `nuxt.config` :

````js
  /**
   * Auth module configuration
   */
  auth: {
    redirect: {
      login: '/login',
      callback: '/callback',
      home: '/',
    },
    strategies: {
      auth0:{
        authorization_endpoint: `https://${env.AUTH0_DOMAIN}/authorize`,
        userinfo_endpoint: `https://${env.AUTH0_DOMAIN}/userinfo`,
      },
    },
    plugins: [
      { src: '~/plugins/auth.js' },
    ],
  },
````

However, this causes the endpoints of the `dev` environment to be hardcoded into the `Nuxt` build... which means the wrong endpoints would be used in the `staging` environment.

**Comment by morficus** [comment]
@jslegers for you case specifically, you could setup a proxy to solve this.
that way your auth0 endpoints are just `/auth/authorize` and `/auth/userinfo` (no domain name attached) and the proxy config would then use the `AUTH0_DOMAIN` env var.

Your nuxt.config.js file would look like this
```js
strategies: {
  auth0:{
    authorization_endpoint: '/auth/authorize',
    userinfo_endpoint: '/auth/userinfo',
  },
},

proxy: {
  '/auth': {
    target: env.AUTH0_DOMAIN,
    pathRewrite: {
      '^/auth': '/'
    }
  }
```

**Comment by albanm** [comment]
FYI I worked on a small module to solve difficulties encountered around this subject.

https://github.com/koumoul-dev/nuxt-config-inject

It is a brutal solution based on search and replace in the built files. It seems to be working quite well for our needs right now. Config elements that are interpreted in an advanced way at build times will not work. But env vars, metas, i18n messages, webpack public path, router base, etc all seem to work.

Not satisfactory in many ways, but hey, right now we have containers starting in 1 or 2 seconds and fetching config directly from multiple environments.

**Comment by jslegers** [comment]
> @jslegers for you case specifically, you could setup a proxy to solve this.
> that way your auth0 endpoints are just `/auth/authorize` and `/auth/userinfo` (no domain name attached) and the proxy config would then use the `AUTH0_DOMAIN` env var.

And I suppose the idea is that the proxy would forward to the correct endpoint based on the referrer path?

I guess this could work as a workaround, but I'd rather avoid this extra forwarding. Also, there are other parameters we'd like to keep different between environments.

In fact, I just learnt that we're even using the same build for production... which only further complicates this matter.

**Comment by morficus** [comment]
> And I suppose the idea is that the proxy would forward to the correct endpoint based on the referrer path?

@jslegers I'm not sure what you are calling the "referrer path", but take a look at the [proxy module documentation](https://github.com/nuxt-community/proxy-module) and the config I added should make more sense. I don't mind helping if you have more questions.

**Comment by kensixx** [comment]
> I couldnt agree more, currently the env variables dont really cut it, espically when using docker images etc.
> 
> The reality of using `nuxt-axios` in a docker image, currently using `nuxt-env`
> 
> ```
> export default function({ $axios, app }) {
>   if (process.server) {
>     $axios.defaults.baseURL = `http://localhost:${app.$env.PORT || '5000'}`
>   } else {
>     $axios.defaults.baseURL = app.$env.EP_BASE_URL || `http://localhost:${app.$env.PORT || '5000'}`
>   }
> }
> ```

I'm getting a headache for about 2 days over this problem with Nuxt on my runtime env variables using k8s and Docker..

Hi @blowsie, this really helped editing the `nuxt-axios`'s `plugins/axios.js` using `nuxt-env`. Thank you so much.. 

Hope the Nuxt core team resolves this one out by supporting runtime env variables especially when using Kubenetes, CI/CD, Docker, etc.

**Comment by kensixx** [comment]
> > app.$axios.defaults.baseUrl = app.$env.BASE_URL
> 
> Do you have an example, I just started with NUXT and I'm confused where to set this

this is the whole code of my `plugins/axios.js`, hope it helps (if you still need it)

```
export default function({ $axios, app }) {
  // set Nuxt Axios' global configuration to use this runtime env variable to make requests
  // Resource: https://github.com/nuxt/nuxt.js/issues/5100#issuecomment-587337547
  // Resource: https://github.com/samtgarson/nuxt-env
  $axios.defaults.baseURL = app.$env.API_URL

  $axios.onRequest((config) => {
    // Adds this header to all requests
    config.headers['Access-Control-Allow-Origin'] = config.baseURL
    config.timeout = 1800000
  })
}
```

**Comment by hynding** [comment]
Can someone identify where in webpack the process.env variables are being replaced during the build process? I'm assuming the "build" nuxt.config option has already been explored for preventing this functionality (and failed).

**Comment by blowsie** [comment]
@hynding , its simply when the server file is generated its taking the variables from the node runtime

**Comment by rick-stevens** [comment]
Our company uses multi-stage deployment as well (k8s) where the build step has no access to the runtime environment variables at all. We're often using env in server middleware and `nuxt.config.js` where `nuxt-env` doesn't reach. Been eagerly awaiting support for over a year. Please help!

**Comment by samtgarson** [comment]
@rick-stevens surely in nuxt config and server middleware you can just use `process.env`? These files are only run server side AFAIK

**Comment by samtgarson** [comment]
I misread your bit about the multistage build 🤦‍♂️😅

**Comment by chanon** [comment]
I just wasted about 4 hours trying to get my Nuxt app to run in my Kubernetes cluster because of this.

Whatever the solution, please make a clear distinction between 
- environment on server when app was built (eg. when use nuxt build or nuxt generate)
- environment on server when app was run (eg. when use nuxt start)
Maybe make it app.$runtimeEnv and app.$buildEnv 

Configuration could be similar to nuxt-env, but separate them so you can set which env vars you want from runtime and from buildtime. Separating them like this will make it clear where the env is coming from.

**Comment by DreaMinder** [comment]
I've created a module based on this implementation  https://github.com/nuxt/nuxt.js/issues/5100#issuecomment-476032241

https://github.com/DreaMinder/nuxt-env-injector
⚠️ Don't use It in production yet. ⚠️

I would appreciate any kind of feedback. In case it is not what you expect as a solution to this issue.

**Comment by mefcorvi** [comment]
@DreaMinder You may want to take a look at https://github.com/koumoul-dev/nuxt-config-inject, it handles several edge cases that aren't supported in your solution. For example, in some cases it's important whether value starts with `http://`or not, you can't use just `%%INJECTED_myvar%%` there.

**Comment by JulianMar** [comment]
There is a different approach to this problem.

What if all your envs are actually stored in vuex. Like [this](https://dev.to/mrnaif2018/how-to-load-dynamical-environment-variables-with-nuxt-js-j7k) blog post suggests. With this approach you can use runtime envs for server stuff and use vuex to access the envs on the client. 
We tried this approach and it worked really well.

Please let me know if someone sees an issue with using this approach

**Comment by blowsie** [comment]
@JulianMar 
Sometimes, or alot of the time you only want the environment variables to be available on the serverside, and sometimes you want to use them before vuex kicks in.

**Comment by JulianMar** [comment]
so use just runtime envs and don't add them to the `env` property in nuxt, this way the won't get replaced. 

For the one you need just push them to the frontend with vuex. If you need something before vuex is initialized you have to use `process.env.VAR_NAME`

**Comment by DreaMinder** [comment]
@JulianMar If you take a look at the first message, you will see nuxt-env module mentioned there. Its approach very similar to vuex-approach. So basically this whole topic started because this approach doesn't work well.
Here is an example of a case when it doesn't work: I need baseURL for the axios-module to be changeable. I only use env. variable in nuxt.config to set baseURL, there is no way vuex would help in this case.

**Comment by samtgarson** [comment]
The Vuex approach works fine but IMO application configuration and application state are separate things and deserve separate toolsets. 

Let’s be clear here—there is _no_ approach that would fix this issue with modules like axios-module, without a change to those module themselves. 

It’s perfectly possible for axios-module and others to use an approach like nuxt-env (or another) to read configuration, it just needs to change from writing its configuration to file at build time. 

What we need is a direction or decision from @pi0, @manniL, @alexchopin or @Atinux on how they want Nuxt to become 12-factor compliant, so that module authors will fall in line.

**Comment by nachogarcia** [comment]
Totally agree @samtgarson .

Currently some of those modules allow to change the configuration in a plugin (you can defer init on GTM and init on a plugin, set the configs on axios on a plugin...) but it's not the convention.

Right now the standard is to pass the env variables in nux.config.js and it is clearly not enough.

There are some modules that help a bit like we mentioned before but I don't think that _installing up to 2 modules to have default env variables that can change in runtime except the ones in nuxt.config_ or going full DIY is the way to go for something that should be provided out of the box and standarized for all modules

**Comment by JulianMar** [comment]
What about just removing the [define](https://webpack.js.org/plugins/define-plugin/) webpack plugin on the server side? 
After some investigation I found [this](https://github.com/nuxt/nuxt.js/blob/85383688371f5ba7b079e1d829a1ada91bc11eaf/packages/webpack/src/config/base.js#L150) code line in the webpack config.

Maybe this function can be moved into the client compilation. By moving it, the envs in the server script shouldn't get replaced with strings.

**Comment by IGassmann** [comment]
It's frustrating to not receive any direction or a mere update on this thread by the Nuxt's core developers. More communication would definitely be appreciated here.

**Comment by Atinux** [comment]
We are aware of this and are looking for a nice solution to implement it right inside Nuxt ☺️

**Comment by tommcclean** [comment]
#metoo, our build generates an artefact used for 7 different environments.

**Comment by mdkhwajams** [comment]
https://webpack.js.org/configuration/optimization/#optimizationnodeenv

**Comment by pi0** [comment]
#7312 Hopefully landing with 2.13 🚀

**Comment by samtgarson** [comment]
Amazing news, thanks @pi0! PR looks great. 

Do you know if there are any plans to help module authors in the community (e.g. axios module) to adopt these new configuration options? That would be the last step to help get these issues sorted for good.

**Comment by zhangpengchen** [comment]
> #7312 Hopefully landing with 2.13 🚀

Amazing news, thanks @pi0, when will you expect to release 2.13. There is no preview/nightly build we could try this.

**Comment by manniL** [comment]
> > #7312 Hopefully landing with 2.13 🚀
> 
> Amazing news, thanks @pi0, when will you expect to release 2.13. There is no preview/nightly build we could try this.

There is, it's called `nuxt-edge`. https://www.npmjs.com/package/nuxt-edge

**Comment by zouyn** [comment]
look forward to 2.13

**Comment by joaquimds** [comment]
From what I've found, `nuxt-env` doesn't handle SPA mode – it uses a server-side plugin to work, which doesn't get executed in SPA mode.

However, I have had success with this:

`app.html`
---
```
<!DOCTYPE html>
<html {{ HTML_ATTRS }}>
<head {{ HEAD_ATTRS }}>
  {{ HEAD }}
</head>
<body {{ BODY_ATTRS }}>
<script>
  window.env = {{ JSON.stringify(ENV) }}
</script>
{{ APP }}
</body>
</html>
```

That `ENV` template variable holds the same variables as the Nuxt `env` configuration parameter, and also includes `NUXT_ENV_` variables automatically.

If this is a really bad idea, please let me know!

**Comment by pi0** [comment]
Hi @joaquimds you can now try [RuntimeConfig](https://nuxtjs.org/guide/runtime-config) which also supports SPA (no need to inject from app.html)

Closing the issue as it is landed to 2.13.0.

Thanks again @janesser @samtgarson and all feedbacks on this issue :heart:

**Comment by blowsie** [comment]
@pi0 [RuntimeConfig](https://nuxtjs.org/guide/runtime-config) is great, but it doesnt allow use to use environment variables in our `nuxt.config.js` in a dynamic way. Think different configuration think **Test/Staging/Prod**

Is there an open ticket for this problem? Should this one be re-opened?

**Comment by SkyaTura** [comment]
🤔  I'm not sure if I got your point... could you elaborate it?

_(I guess another issue would be better, since the main subject of this one was in fact solved)_

**Comment by cageyv** [comment]
@blowsie i used a dirty sed solution in my last project with old nuxt.js version 2.11.0. 
I never try nuxt after 2.13 release, but someone recommend me nuxt-edge package for this problem.
So if you have time please check with nuxt-edge

**Comment by blowsie** [comment]
@cageyv , im already using 2.13 with the runtime config, but it doesnt help the fact when you use  `process.env`  in your `nuxt.config` the variables are translated to strings at build time.

@SkyaTura does this answer your question?

**Comment by blowsie** [comment]
> nother issue would be better, since the main subject of this one was in fact solved

Sadly this is not true, the main subject of this was immutable build artifcats, we dont have that yet

**Comment by samtgarson** [comment]
@blowsie are you using SPA mode or Universal mode? 

I'm not as familiar with SPA mode enough to suggest a fix for this, but if you're using Universal mode (i.e. SSR) then using `process.env` should be run at runtime as well—the only issue from here are 3rd party modules which write their configuration to file at build time (like the Axios module).

For these, I think it would be great if the Nuxt team could help nudge the major in the right direction to support runtime config, I know when people used `Nuxt-env` this was a common frustration.

**Comment by blowsie** [comment]
@samtgarson I am indeed using universal mode. 

But you are right, the issues im having are with the nuxt modules, like for example nuxt-community/auth-module https://github.com/nuxt-community/auth-module/issues/713#issuecomment-676098189

**Comment by samtgarson** [comment]
Yep. I don't think there's much Nuxt can do while those modules are writing config to file other than help module authors migrate to this strategy.

**Comment by clairegraham** [comment]
I just ran into this issue.

My use case:
Our production site uses a different Google Tag Manager ID from the staging site. So, I still deploy the site to both environments in a "production" node environment, but they have various configuration differences based on where they end up being deployed.

I tried to use nuxt-env (which is a great module!) but it wasn't accessible via nuxt.config.js, which is where the GTM module configuration resides. This was my solution:

```package.json```:
```
{
  "scripts": {
    "deploy:staging": "DEPLOY_TARGET=staging nuxt build"
    "deploy:prod": "DEPLOY_TARGET=production nuxt build"
  }
}
```

```nuxt.config.js```
```
module.exports = {
  gtm: {
    id: process.env.DEPLOY_TARGET == 'staging' ? 'GTM-STAGING-ID' : 'GTM-PROD-ID'
  }
}
```

**Comment by pi0** [comment]
Hi @blowsie indeed as @samtgarson mentioned, it is up to module authors supporting runtime config which auth module still hasn't. 

@clairegraham You can try latest version of gtm-module. It supports runtime config: https://github.com/nuxt-community/gtm-module#runtime-config

```js
export default {
  modules: [
    '@nuxtjs/gtm'
  ],

  publicRuntimeConfig: {
    gtm: {
      id: process.env.DEPLOY_TARGET == 'staging' ? 'GTM-STAGING-ID' : 'GTM-PROD-ID'
    }
  },
}
```

**Comment by ababbdev** [comment]
Hi,
Thank you for doing the work to support the dynamic env values! 
We have been migrating to use this strategy however I just want to confirm if this will be the strategy that is carried over to the next version of Nuxt when Vue 3 is added or these are up for changes?

**Comment by memic84** [comment]
Doesn't this solve your problem?
https://nuxtjs.org/blog/moving-from-nuxtjs-dotenv-to-runtime-config
https://nuxtjs.org/docs/2.x/configuration-glossary/configuration-runtime-config

**Comment by rotemrevivo91** [comment]
No, unfortunately. The problem is that i have a single build (i.e 1 docker image) that is used on multiple environments. Each user should be able to specify their own router base as an .env variable and just run the container without the need to rebuild the image.
Nuxt resources are then can't be found (nuxt .js files).
I was able to use:
nuxt.config.js
```
  hooks: {
    render: {
      resourcesLoaded(resources) {
        const path = `${process.env.BASE_URL_SUB_DIRECTORY}_nuxt/`
        resources.clientManifest && (resources.clientManifest.publicPath = path)
        resources.modernManifest && (resources.modernManifest.publicPath = path)
        resources.serverManifest && (resources.serverManifest.publicPath = path)
      }
    }
  }
```
plugins\cdn.js:
```
export default function({$config}) {
  if (process.client) {
    __webpack_public_path__ = `${$config.baseURLSubDirectory}_nuxt/`
  }
}
```
which replaces all resources on run-time! which is important but still i get a 404 page even though nuxt is able to find all the needed resources.

**Comment by pi0** [comment]
@rotemrevivo91 Would you please open a feature request for supporting runtime baseURL?

**Comment by rotemrevivo91** [comment]
@pi0 will do :)

**Comment by alexcroox** [comment]
Did anyone find a solution to supporting multiple environments? The Nuxt docs seem to assume that everyone is running production locally or at best only has dev + production. If you have local/sandbox/staging/production this runtime config doesn't seem suitable for large applications with multiple environments before production.

**Comment by samtgarson** [comment]
I've used it to run many dynamic environments (we ran a hand-rolled version of Vercel'a preview environments with a deployment for each branch as well as production). 

I'd suggest you open a Discussion on the nuxt repo with your specific issues, as (apart from specific pieces of config like above) runtime config should allow you to have different config values per environment. 

As mentioned above, many community modules still need to adopt this feature 😕

**Comment by rotemrevivo91** [comment]
We opened a new feature request https://github.com/nuxt/nuxt.js/issues/8509#issuecomment-747453183
The nuxt team will work on way to make it easier but we also found a work around. If you need extra help don't hesitate to ask :)

**Comment by pi0** [comment]
Locking issue to avoid confusions. Runtime config for multiple envs is supported. Requested feature is for setting baseURL (router.base) after build.
